code/data_dictionary.R

```diff
@@ -46,27 +46,62 @@ adbs <- adbs %>%
+uszips <- read.csv("./data/simplemaps_uszips_basicv1.72/uszips.csv")
 
 length(adbs$Zip[(!adbs$Zip %in% sprintf("%05d", uszips$zip))])            # 42,580 rows without a zip in the list
 length(unique(adbs$Zip[(!adbs$Zip %in% sprintf("%05d", uszips$zip))]))    # 5,036 unique zips
 sample(unique(adbs$Zip[(!adbs$Zip %in% sprintf("%05d", uszips$zip))]), 5) # this tends to give what are, according to Google Maps, valid ZIPs. Perhaps the simplemaps list is too old...
 # we do not appear to be much closer, but at least the error rate isn't horrible (42,000 out of 4,800,000)
 
+#Data Validation: Zip --------------------------------------------------------------
+#True indicates entry is a valid zip, i.e. it exists in the zip-reference list 
+adbs <- adbs %>% mutate(ValidZip = case_when(
+  is.na(Zip) ~ "No Entry",
+  Zip %in% sprintf("%05d", uszips$zip) ~ "True", 
+  !Zip %in% sprintf("%05d", uszips$zip) ~ "Fail"))
+
+table(adbs$ValidZip, useNA = "always") 
+
+
 ### Data Check: State Names -----------------------------------------------
 # check against US Census data: American National Standards Institute (ANSI) Codes for States, the District of Columbia, Puerto Rico, and the Insular Areas of the United States
 #via: https://www.census.gov/library/reference/code-lists/ansi.html
-statecodes <- read.table("../data/census/state.txt", sep = "|", header = TRUE)
+statecodes <- read.table("./data/census/state.txt", sep = "|", header = TRUE)
 
 table(adbs$State[(!adbs$State %in% statecodes$STUSAB)]) # list counts of unmatched results, showing 1 AE, 1 FI, 210 XX
 
 adbs[adbs$State == "FI",] # based on the ZIP code, this should be FL, and can be 'fixed' easily enough as part of final data cleaning
 adbs[adbs$State == "AE",] # when viewing ZIP, Lending data, this does indeed appear to be tied to a military address in Europe, Middle East, Africa or Canada
 
+# Impute Missing Fields: state --------------------------------------------------------------
+# Impute missing state by zipcode. Here, we do not check if city and zipcode match
+cat(sprintf("%s missing states", nrow(adbs[adbs$ImputedState=='XX',])))
+XXstate_zip = adbs[(adbs$State=='XX' & !is.na(adbs$Zip)),]$Zip
+XXstate_city = adbs[(adbs$State=='XX' & !is.na(adbs$Zip)),]$City
+cat(sprintf("%s missing states, with zip entry", length(XXstate_zips)))
+
+
+adbs$ImputedState =adbs$State
+imputed_states = c()
+missing_zips = c()
+for (zip in XXstate_zip){
+  state = uszips[uszips$zip==zip,]$state_id
+  
+  if (identical(state, character(0))){
+    state = NA
+    missing_zips = c(missing_zips, zip)}
+  imputed_states=c(imputed_states, state)
+}
+
+adbs[which(adbs$ImputedState=='XX' & !is.na(adbs$Zip)), arr.ind=TRUE]$ImputedState = imputed_states
+
+cat(sprintf("out of the %s missing states that had zip entry, %s states imputed from zipcodes, %s zipcodes not found in 'uszips' file", length(impute_zips), (length(imputed_states) - length(missing_zips)),length(missing_zips)))
```

Comment:
this line does not run: cannot find `impute_zips`

Reply:
Appears fixed.